bindings/rust/src/lib.rs

```diff
@@ -289,6 +289,73 @@ impl TypedExecutionResult {
 }
 
 impl Instance {
+    /// Ensure the range is valid according to the currently available memory size.
+    fn checked_memory_range(
+        &self,
+        offset: u32,
+        size: usize,
+    ) -> Result<core::ops::Range<usize>, ()> {
+        // This should be safe given usize::BITS >= u32::BITS, see https://doc.rust-lang.org/std/primitive.usize.html.
+        let offset = offset as usize;
+        let memory_size = self.memory_size();
+        if memory_size == 0 || offset > memory_size || (offset + size) > memory_size {
```

Comment:
This would allow  empty slices `memory_size..memory_size`.

Should this better be
```suggestion
        if offset >= memory_size || (offset + size) > memory_size {
```
(`memory_size == 0` check would be redundant then)

Reply:
The `memory_size != 0` check is there so that we can have the `debug_assert` for ptr not being null.

Reply:
> This would allow empty slices memory_size..memory_size.

Also see https://github.com/wasmx/fizzy/pull/609#discussion_r571211243

Reply:
My comment was not about empty slice restriction, but another restriction to allow slices to start only at valid offset.

Reply:
Well, empty slice at the very end. Well, the `offset + size > memory_size` prevents generating a slice which would do any overread, so `65536, 0` as an input right now is accepted. See tests.

Reply:
Oh, I should have also used tests with 65535.

Reply:
I suspect now that it was better not to listen to me and leave only the check `if (offset + size) > memory_size {err}` (i.e. allow `memory_size..memory_size` empty slices) 

I can't be sure about Rust, but in C/C++ it's allowed to have a range of pointers/iterators with both ends pointing to one-past-last element.